dali/pipeline/operators/fused/crop_mirror_normalize.cc

```diff
@@ -56,92 +56,70 @@ normalization only.
-  mirror_[ws->data_idx()] = spec_.GetArgument<int>("mirror", ws, ws->data_idx());
+bool CropMirrorNormalize<CPUBackend>::InferOutputs(std::vector<kernels::TensorListShape<>> &shape,
+                                                   std::vector<TypeInfo> &type,
+                                                   const HostWorkspace &ws) {
+  shape.resize(1);
+  type.resize(1);
+  CommonSetup(ws);
+  const auto &input = ws.InputRef<CPUBackend>(0);
+  auto &output = ws.OutputRef<CPUBackend>(0);
+  std::size_t number_of_dims = input.shape().sample_dim();
+  DALI_TYPE_SWITCH_WITH_FP16_CPU(input_type_, InputType,
+    DALI_TYPE_SWITCH_WITH_FP16_CPU(output_type_, OutputType,
+      VALUE_SWITCH(number_of_dims, Dims, (3, 4),
+      (
+        using Kernel = kernels::SliceFlipNormalizePermuteCPU<OutputType, InputType, Dims>;
+        using Args = kernels::SliceFlipNormalizePermuteArgs<Dims>;
+        type[0] = TypeInfo::Create<OutputType>();
+        shape[0].resize(batch_size_, Dims);
+        kmgr_.Initialize<Kernel>();
+        // Do the kernel setup
+        for (int sample_idx = 0; sample_idx < batch_size_; sample_idx++) {
+          auto in_view = view<const InputType, Dims>(input[sample_idx]);
+          auto &kernel = kmgr_.Get<Kernel>(sample_idx);
+          kernels::KernelContext ctx;
+          auto req = kernel.Setup(ctx, in_view, any_cast<Args&>(kernel_sample_args_[sample_idx]));
+          shape[0].set_tensor_shape(sample_idx, req.output_shapes[0][0].shape);
+        }
+      ),
+      (
+        DALI_FAIL("Not supported number of dimensions: " + std::to_string(number_of_dims));
+      ));
+    )
+  );
 
-  auto &output = ws->Output<CPUBackend>(0);
-  output.SetLayout(output_layout_);
+  return true;
 }
 
 template <>
 void CropMirrorNormalize<CPUBackend>::RunImpl(SampleWorkspace *ws) {
-  this->DataDependentSetup(ws);
   const auto &input = ws->Input<CPUBackend>(0);
   auto &output = ws->Output<CPUBackend>(0);
   auto data_idx = ws->data_idx();
+  auto sample_idx = data_idx;
+  std::size_t number_of_dims = input.shape().sample_dim();
 
   DALI_TYPE_SWITCH_WITH_FP16_CPU(input_type_, InputType,
     DALI_TYPE_SWITCH_WITH_FP16_CPU(output_type_, OutputType,
-      detail::RunHelper<OutputType, InputType>(
-        output, input, slice_anchors_[data_idx], slice_shapes_[data_idx],
-        mirror_[data_idx], pad_output_, mean_vec_, inv_std_vec_);
+      VALUE_SWITCH(number_of_dims, Dims, (3, 4),
+      (
+        using Kernel = kernels::SliceFlipNormalizePermuteCPU<OutputType, InputType, Dims>;
+        using Args = kernels::SliceFlipNormalizePermuteArgs<Dims>;
+        auto &kernel = kmgr_.Get<Kernel>(sample_idx);
+        auto in_view = view<const InputType, Dims>(input);
+        auto out_view = view<OutputType, Dims>(output);
+        auto &args = any_cast<Args&>(kernel_sample_args_[sample_idx]);
+        kernels::KernelContext ctx;
+        kernel.Run(ctx, out_view, in_view, args);
```

Comment:
Formally, this invocation is incorrect - you don't respect scratch requirements nor provide a scratchpad. This introduces dependence of this operator on a particular implementation of the kernel.